code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/integration/ldap/LdapIdentitySearchProvider.java

```diff
@@ -0,0 +1,381 @@
+            return username;
+        } else {
+            return LdapConstants.LDAP_LOGIN_DOMAIN.get() + '\\' +username;
+        }
+    }
+
+    public String getSamName(String username) {
+        if (!isConfigured()) {
+            return null;
+        }
+        if (username.contains("\\")) {
+            return username.split("\\\\", 2)[1];
+        } else {
+            return username;
+        }
+    }
+
+    private List<Identity> searchGroup(String name, boolean exactMatch) {
+        LdapContext context = getServiceContext();
+        String query;
+        if (exactMatch) {
+            query = "(&(" + LdapConstants.GROUP_SEARCH_FIELD.get() + '=' + name + ")(" + LdapConstants.OBJECT_CLASS + '='
+                    + LdapConstants.GROUP_OBJECT_CLASS.get() + "))";
+        } else {
+            query = "(&(" + LdapConstants.GROUP_SEARCH_FIELD.get() + "=*" + name + "*)(" + LdapConstants.OBJECT_CLASS + '='
+                    + LdapConstants.GROUP_OBJECT_CLASS.get() + "))";
+        }
+        return searchLdap(context, LdapConstants.LDAP_DOMAIN.get(), name, query);
+    }
+
+    private List<Identity> searchUser(String name, boolean exactMatch) {
+        LdapContext context = getServiceContext();
+        String query;
+        if (exactMatch)
+        {
+            query = "(&(" + LdapConstants.USER_SEARCH_FIELD.get() + '=' + name + ")(" + LdapConstants.OBJECT_CLASS + '='
+                    + LdapConstants.USER_OBJECT_CLASS.get() + "))";
+        } else {
+            query = "(&(" + LdapConstants.USER_SEARCH_FIELD.get() + "=*" + name + "*)(" + LdapConstants.OBJECT_CLASS + '='
+                    + LdapConstants.USER_OBJECT_CLASS.get() + "))";
+        }
+        return searchLdap(context, LdapConstants.LDAP_DOMAIN.get(), name, query);
+    }
+
+    private Identity getObject(String distinguishedName, String scope) {
+        try {
+            LdapContext context = getServiceContext();
+            if (context == null) {
+                return null;
+            }
+            Attributes search = context.getAttributes(new LdapName(distinguishedName));
+            if (!isType(search, scope) && !hasPermission(search)){
+                return null;
+            }
+            return attributesToIdentity(search);
+        } catch (NamingException e) {
+            logger.error("Failed to get object: " + distinguishedName, e);
+            return null;
+        }
+    }
+
+    private boolean isType(Attributes search, String type) throws NamingException {
+        NamingEnumeration<?> objectClass = search.get(LdapConstants.OBJECT_CLASS).getAll();
+        boolean isType = false;
+        while (objectClass.hasMoreElements()) {
+            Object object = objectClass.next();
+            if ((object.toString()).equalsIgnoreCase(type)){
+                isType = true;
+            }
+        }
+        return isType;
+    }
+
+    private LdapContext getServiceContext() {
+        if (isConfigured()) {
```

Comment:
Everywhere this is called, you've checked isConfigured higher up the stack. Is it redundant to do it again here?